src/librustc_passes/rvalue_promotion.rs

```diff
@@ -364,26 +412,24 @@ fn check_expr<'a, 'tcx>(
                 Def::AssociatedConst(did) => {
                     let promotable = if v.tcx.trait_of_item(did).is_some() {
                         // Don't peek inside trait associated constants.
-                        false
+                        NotPromotable
+                    } else if v.tcx.at(e.span).const_is_rvalue_promotable_to_static(did) {
+                        Promotable
                     } else {
-                        v.tcx.at(e.span).const_is_rvalue_promotable_to_static(did)
+                        NotPromotable
                     };
-
                     // Just in case the type is more specific than the definition,
                     // e.g. impl associated const with type parameters, check it.
                     // Also, trait associated consts are relaxed by this.
-                    v.promotable &= promotable || v.type_has_only_promotable_values(node_ty);
-                }
-
-                _ => {
-                    v.promotable = false;
+                    promotable | v.type_promotability(node_ty)
```

Comment:
This is the only place `|` is used, since misusable. Also, the method was named `type_has_only_promotable_values` before, which suggests it's impossible to create an unpromotable value of that type. `type_promotability` is a bit more general but it does highlight a point: this code is probably wrong.

@oli-obk this should probably not be relaxed like this. You might need to reintroduced bitflags to indicate the *reason* for unpromotability Q_Q. (i.e. you can remove `INTERIOR_MUT` and `NEEDS_DROP` flags based on the type but you can't remove anything pertaining to "creates values that cannot be operated on correctly").

Reply:
I have a feeling we discussed this before and it ended up being a breaking change to change this.

"has only promotable values" made sense pre-miri. Not so much afterwards.

This is pretty much the topic I'm addressing in https://github.com/rust-lang/rust/pull/51570